### PR TITLE
add docs and comments to supercatchup

### PR DESCRIPTION
**Pull request by QuiteStochastic**
add comments to supercatchup.ml, as well as reformat and add better log statements.  all code changes should be trivial, nothing but log statements have been changed.

### CODE REVIEW

**Review by lk86 — APPROVED** [review]

**Review by mimoo — APPROVED** [review]

File: src/lib/ledger_catchup/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+- Breadcrumb: a breadcrumb is a block which has been fully validated and given an attached and connected ledger mask which connects to the previous mask which eventually connects to a ledger, and so an attached mask in essence is a full ledger. breadcrumbs are the "nodes" in a transition frontier. breadcrumbs can only exist if the parent breadcrumb already exists in the transition frontier, because the parent is necessary for validating and of course a fully attached ledger mask can only exist if there exists a previous ledger.  the transition frontier, being made of breadcrumbs, is therefore always fully connected.
+
+- Transition frontier: the transition frontier is a tree of breadcrumbs with max height of `k`, where `k` is a constant defined by the crypto team, currently set at `k = 209`.  the transition frontier is the very most recent `k` number of blocks, as well as all the forks, or sidechains, off of the current longest chain.  the block-chain path through the transition frontier with the best chain quality metric is considered the current canonical chain, and the tip of this canonical chain is called the best-tip.
```
Comment (nholland94) [inline src/lib/ledger_catchup/README.md:9]: Correction: `k = 290`

File: src/lib/ledger_catchup/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+### Quick definitions
+
+- Block recap: a block contains a hash of the current ledger, and a `staged_ledger_diff`, which is effectively a list of transactions that change the previous staged ledger to the current block's staged ledger. A transaction is of the form "acct a sends X mina tokens to acct B". It doesn't tell you the balances of acct A and B. The block does not contain the ledger in it's entirety, nor can any account balances be looked up.  https://docs.minaprotocol.com/en/architecture/whats-in-a-block#staged-ledger-diff
+
+- Breadcrumb: a breadcrumb is a block which has been fully validated and given an attached and connected ledger mask which connects to the previous mask which eventually connects to a ledger, and so an attached mask in essence is a full ledger. breadcrumbs are the "nodes" in a transition frontier. breadcrumbs can only exist if the parent breadcrumb already exists in the transition frontier, because the parent is necessary for validating and of course a fully attached ledger mask can only exist if there exists a previous ledger.  the transition frontier, being made of breadcrumbs, is therefore always fully connected.
```
Comment (nholland94) [inline src/lib/ledger_catchup/README.md:7]: There are some capitalization mistakes throughout (eg here: `... previous ledger.  the transition frontier ...`)

File: src/lib/ledger_catchup/README.md (hunk 1 above)
Comment (nholland94) [reply src/lib/ledger_catchup/README.md:9]: >  the block-chain path through the transition frontier with the best chain quality metric is considered the current canonical chain

I would rephrase this as "the strongest chain" rather than the "best chain quality metrics" (chain quality is one of the factors that determines strength, and strength is what we actually compare when determining the best path).

File: src/lib/ledger_catchup/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+- Transition frontier: the transition frontier is a tree of breadcrumbs with max height of `k`, where `k` is a constant defined by the crypto team, currently set at `k = 209`.  the transition frontier is the very most recent `k` number of blocks, as well as all the forks, or sidechains, off of the current longest chain.  the block-chain path through the transition frontier with the best chain quality metric is considered the current canonical chain, and the tip of this canonical chain is called the best-tip.  
+
+- Bootstrap, high level: bootstrap is the process of using zk-snarks to validate the validity of the chain, from the last known block (which would be the genesis block for fresh daemons just starting up) to the best tip.
```
Comment (nholland94) [inline src/lib/ledger_catchup/README.md:11]: A better description would be "bootstrap is the process by which nodes join the network, synchronizing to the network's best tip".

File: src/lib/ledger_catchup/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    - the catchup tree is a trees of hashes of blocks, which are in a pending state and need to be downloaded and processed into the transition frontier. (it's actually a forest and not one tree, but this is just because the scheduler passes to catchup a bunch of blocks in a tree minus their common root, so it becomes a forest)
+- within the `run` function of super catchup, first it sets up a number of dependencies and creates a few aux functions.
+    - there is a very important interior function defined within the super_catchup run function called `run_node`. `run_node` is a state machine which puts each node (ie a job) of the catchup tree through it's phases. it takes a node of the transition frontier, and either fully downloads it, validates it, builds it into a breadcrumb, or builds the parent into a breadcrumb. the function looks like a big match-case expression that handles all the stages of a node. it recursively calls itself to handle the subsequent stages.
```
Comment (nholland94) [inline src/lib/ledger_catchup/README.md:24]: Might be good to expand on this some more. For each node we want to perform catchup on, we call `run_node` on it. This function waits for conditions and triggers actions before recursively invoking itself to enact a state transition on that node, until the function eventually exits when the node is done being processed (or fails).

File: src/lib/ledger_catchup/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    - the downloader is also created in the beginning of the `run` function. has a list of blocks that need to be downloaded. super catchup (i believe) downloads in a more forward order towards the target.
```
Comment (nholland94) [inline src/lib/ledger_catchup/README.md:25]: Yes, the downloader is designed to download blocks in the same order they were produced in (as opposed to normal catchup, which did it backward).

File: src/lib/ledger_catchup/super_catchup.ml [hunk 6]
```diff
@@ -1077,167 +1081,180 @@ let run_catchup ~logger ~trust_system ~verifier ~network ~frontier ~build_func
...
+                ~f:
+                  (Rose_tree.iter ~f:(fun c ->
                        let node =
-                         create_node t ~downloader (`Hash (h, l, parent))
+                         create_node ~downloader t (`Initial_validated c)
                        in
-                       don't_wait_for (run_state_machine node >>| ignore) ) ;
-                     (h, l))
-                 : State_hash.t * Length.t )))
+                       ignore
```
Comment (nholland94) [inline src/lib/ledger_catchup/super_catchup.ml:1243]: don't_wait_for is generally preferred over ignore in this case, because it provides more direction about the purpose of the ignore call (they are functionally the same though).

**Review by nholland94 — APPROVED** [review]
